### PR TITLE
feat: use cloud storage from amalthea

### PR DESCRIPTION
**Pull request by olevski**
Adds cloud storage for the new amalthea operator and with the new csi rclone version.

/deploy renku=feat-jupyter-free-sessions amalthea-sessions=feat-add-cloud-storage renku-notebooks=feat-jupyter-free-sessions renku-gateway=feat-add-renku-refresh-token-for-data-service renku-ui=andrea/custom-launcher-jupyter-free extra-values=amalthea-sessions.deployCrd=false,amalthea-sessions.rcloneStorageClass=csi-rclone-387-secret-annotation #notest

### CODE REVIEW

**Comment by olevski** [comment]
This change is part of the following stack:

- #360
    - #338
        - #375
            - #387 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: components/renku_data_services/notebooks/blueprints.py
Comment (Panaetius) [inline components/renku_data_services/notebooks/blueprints.py:844]: question: historically, users could mount arbitrary, unsaved cloud storage and the UI would just post a body here, in which case it wouldn't have a storage_id. I don't think this is around in v2 but v1 might still use this. So I'm not entirely sure if this if statement will break things and we'd have to generate ad-hoc ids or if this is fine?

File: components/renku_data_services/notebooks/blueprints.py
Comment (Panaetius) [inline components/renku_data_services/notebooks/blueprints.py:874]: nitpick: spelling

File: components/renku_data_services/notebooks/crs.py
Comment (Panaetius) [inline components/renku_data_services/notebooks/crs.py:36]: question: `SecretRef1` seems like a pretty bad name in the originating file, could this be changed or is this something autogenerated that we can't do anything about?

**Review by Panaetius — CHANGES_REQUESTED** [review]
General Question: there's still https://github.com/SwissDataScienceCenter/renku-data-services/pull/347 that does a bit of cleanup in cloud storage and probably will cause quite a bit of conflict with this, so I'm wondering what the best way would be for us to bring this together with your changes. There's also lots of changes @leafty is working on in regards to cloud storage, and 3 people working on the same bits of code at the same time is bound to cause issues. maybe we should sync?

File: components/renku_data_services/notebooks/crs.py
Comment (olevski) [reply components/renku_data_services/notebooks/crs.py:36]: It is auto-generated. Everything in `cr_jupyter_server.py` and `cr_amalthea_session.py` is autogenerated. That is why I import from these two files into `crs.py` and I rename there. There is no way to change the names in the autogenerated code.

Comment (Panaetius) [reply components/renku_data_services/notebooks/crs.py:36]: can it be changed on the amalthea side where it comes from?

Comment (olevski) [reply components/renku_data_services/notebooks/crs.py:36]: > can it be changed on the amalthea side where it comes from?

No I mean I would have to change the CRD spec for amalthea sessions just so I have a nice name when I autogenerate code. That does not make sense. And the reason there is multiple is that the code generation does not deduplicate all things. And there are multiple `SecretRef` fields in the CRD spec. So TLDR No.

Comment (Panaetius) [reply components/renku_data_services/notebooks/crs.py:36]: ok, makes sense

**Comment by olevski** [comment]
> General Question: there's still https://github.com/SwissDataScienceCenter/renku-data-services/pull/347 that does a bit of cleanup in cloud storage and probably will cause quite a bit of conflict with this, so I'm wondering what the best way would be for us to bring this together with your changes. There's also lots of changes @leafty is working on in regards to cloud storage, and 3 people working on the same bits of code at the same time is bound to cause issues. maybe we should sync?

When it comes to #347 I think we should split out the API breaking change and merge that at some other time. There is still a lot of testing and minor fixes that are needed on the UI as well. So if we can not have the API breaking change then we dont have to distract the UI with these changes (in addition to many already introduced with the new amalthea). You opened #347 @Panaetius do you think you can split out the API breaking change?

As for the cloud storage changes that @leafty will introduce. I think she is standing by until we merge more things. So after we get this PR and the one furtehr in the stack (#411) I think that @leafty can start targeting the same release branch we have been targeting. 

But yeah just to be safe lets have a quick chat in the afternoon tomorrow.

File: components/renku_data_services/notebooks/blueprints.py
Comment (olevski) [reply components/renku_data_services/notebooks/blueprints.py:838]: For this first implementation of the cloud storage in V2 lets keep things simple and not allow people to mount cloud storage that is unsaved. Then if needed we can add this functionality. This also needs some design because in theory we could allow this in many other places in the API where we currently expect something that exists in the DB. So I will make the storage ID required in the post request to launch a session. If a payload is provided then it just overrides the cloud storage saved in our DB.

**Comment by leafty** [comment]
> As for the cloud storage changes that @leafty will introduce. I think she is standing by until we merge more things. So after we get this PR and the one furtehr in the stack (#411) I think that @leafty can start targeting the same release branch we have been targeting.

If I understand things correctly, the new sessions API will not support Renku 1.0. In which case the changes should be reasonable.

**Comment by leafty** [comment]
Note: waiting for a rebase, so I can try to work from this branch.

**Comment by Panaetius** [comment]
> When it comes to https://github.com/SwissDataScienceCenter/renku-data-services/pull/347 I think we should split out the API breaking change and merge that at some other time.

but the API is currently broken and that it works is mostly an accident, any update to sanic or one of its components could break it. So yes, it's a breaking API change but at the same time it's more of a bugfix and unbreaking the API. I wouldn't hold off on this. And it is needed for that PR. Also https://github.com/SwissDataScienceCenter/renku-ui/pull/3299 addressed those changes on the UI AND addresses the changes to storage, so there's a bit of a triangle dependency here.

**Review by Panaetius — APPROVED** [review]

**Pull request merged**
